---
title: "Meningitis Meninges"
subtitle: "Differential Expression: group + sex + GFAP"
author: "Kennedi Todd"
date: "11/07/2023"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---

# Setup
## Working directory
```{r setup}
knitr::opts_knit$set(root.dir = ".")
```

## Libraries
```{r libraries, message=FALSE, warning=FALSE}
library(BiocParallel)       # SnowParam()
library(dplyr)              # left_join()
library(edgeR)              # DGEList()
library(limma)              # plotMDS()
library(ggcorrplot)         # ggcorplot()
library(ggrepel)            # geom_text_repel()
library(ggplot2)            # ggplot()
library(gplots)             # heatmap.2()
library(grDevices)          # colorRampPalette()
library(gridExtra)          # grid.arrange()
library(Matrix)             # Matrix()
library(parallel)           # detectCores()
library(philentropy)        # JSD()
library(rtracklayer)        # import()
library(Seurat)             # GetAssayData()
library(stringr)            # str_match()
library(variancePartition)  # fitExtractVarPartModel()
library(reshape)            # melt()
```

## User defined variables
```{r set_variables}
tissue <- "meninges"
folder <- paste0(tissue,"_1CPM_in_12_samples")
groups <- c("control","meningitis")
group_colors <- c("orange","cornflowerblue")
sex_colors <- c("green","purple")
ann <- "../../refs/chm13v2.0_RefSeq_Liftoff_v5.1.gff3"
model <- "/DEGs"
```

## Save functions
```{r save_functions}
saveToPDF <- function(...) {
    d = dev.copy(pdf,...)
    dev.off(d)
}
```

## Load data
```{r}
dge.filtered.norm <- readRDS(paste0("../../rObjects/",folder,"_DGEList_cell_markers.rds"))
```

# Voom before covariates
## Design matrix
```{r design_matrix}
group <- interaction(dge.filtered.norm$samples$group)
sex <- dge.filtered.norm$samples$sex
gfap <- dge.filtered.norm$samples$GFAP

design <- model.matrix(~ 0 + group + sex + gfap)
design.names <- gsub("group","",colnames(design))
colnames(design) <- design.names
design
```

## Quality weights
```{r voom}
# voom transform counts
v <- voomWithQualityWeights(dge.filtered.norm,
                            design,
                            plot = TRUE)

# save
path <- paste0("../../results/", folder, "/postfiltering_QC/voom_gene_mean_var_weights")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)

# fits linear model for each gene given a series of arrays
fit <- lmFit(v, design)

# contrast design for differential expression
contrasts <- makeContrasts(
  contrasts = c("meningitis - control"),
  levels = design.names)
head(contrasts)

# save contrast names
allComparisons <- colnames(contrasts)
allComparisons # check

# run contrast analysis
vfit <- contrasts.fit(fit, contrasts = contrasts)

# Compute differential expression based on the empirical Bayes moderation of the
# standard errors towards a common value.
veBayesFit <- eBayes(vfit)
plotSA(veBayesFit, main = "Final Model: Mean-variance Trend")

# save
path <- paste0("../../results/", folder, "/postfiltering_QC/voom_gene_final_mean_var")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
```

## Voom MDS Plot
### Sex
```{r mds_prefiltering, warning=FALSE}
  # set colors and get data
sex_color <- sex_colors[v$targets$sex]
names <- v$targets$informative_name

# loop through dimensions and top genes
dims <- list(c(1,2),c(2,3),c(3,4),c(4,5))
topGenes <- c(100,500)

for (dim in dims) {
  
  for (numGenes in topGenes) {
      
    par(bg = 'white')
  
    # plot MDS
    plotMDS(
      v, 
      top = numGenes, 
      labels = names,
      cex = 1, 
      dim.plot = dim, 
      plot = TRUE, 
      col = sex_color
    )
    
    title(paste0("Raw: Top ", numGenes, ", Log2(CPM)"))
    
    #legend(
    #  "topleft",
    #  legend = unique(v$targets$sex),
    #  pch = 16,
    #  col = unique(sex_color),
    #  cex = 1
    #)
    
    # save
    dimTitle <- as.character(dim)
    path <- paste0("../../results/",folder,"/postfiltering_QC/MDS_raw_", 
                   "top_", numGenes, "_",
                   "dim", as.character(dim)[1], "&", as.character(dim[2]), "_colored_by_sex")
    saveToPDF(paste0(path, ".pdf"), width = 8, height = 8)
  }
}

remove(data)
```


### Group
```{r}
# set colors and get data
group_color <- group_colors[v$targets$group]
names <- v$targets$informative_name

# loop through dimensions and top genes
dims <- list(c(1,2),c(2,3),c(3,4),c(4,5))
topGenes <- c(100,500)

for (dim in dims) {
  
  for (numGenes in topGenes) {
    
    par(bg = 'white')
    
    # plot MDS
    plotMDS(
      v, 
      top = numGenes, 
      labels = names,
      cex = 1, 
      dim.plot = dim, 
      plot = TRUE, 
      col = group_color
    )
    
    title(paste0("Raw: Top ", numGenes, ", Log2(CPM)"))
    
    #legend(
    #  "topleft",
    #  legend = unique(v$targets$group),
    #  pch = 16,
    #  col = unique(group_color),
    #  cex = 1
    #)
    
    # save
    dimTitle <- as.character(dim)
    path <- paste0("../../results/",folder,"/postfiltering_QC/MDS_raw_", 
                   "top_", numGenes, "_",
                   "dim", as.character(dim)[1], "&", as.character(dim[2]), "_colored_by_group")
    saveToPDF(paste0(path, ".pdf"), width = 8, height = 8)
  }
}

remove(dims)
```

# Differential Expressioin
## Number of DEGs
Identify number of differentially expressed genes.
```{r decide_tests}
pval <- 0.1

sumTable <- 
  summary(decideTests(
    vfit,  # object
    adjust.method = "BH", # by default the method = "separate"
    p.value = pval,
    lfc = 0  # numeric, minimum absolute log2-fold change required
  ))

print(paste0(tissue, " FDRq < ", pval))
sumTable

write.table(sumTable, 
            paste0("../../results/", folder, model, 
            "/DEG_comparison_FDRq_", format(pval, nsmall = 2),".tsv"), 
            quote = FALSE, 
            sep = "\t")
```

## Output DEG tables
```{r output_DEG_tables}
coef <- 1

for (i in allComparisons) {
  # title 
  title <- gsub(" ", "_", i)
  title <- gsub("-","vs", title)
  
  
  # p < 1, log2fc > 0 ----------------------------------------------------------
  vTopTableAll <-
    topTable(
      veBayesFit, 
      coef = coef,  
      n = Inf, 
      p.value = 1,
      lfc = 0 
    )
  #output as txt file
  path <- paste("../../results/", folder, model, "/", title,"_FDRq_1.00.tsv", sep = "") 
  write.table(
    vTopTableAll,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  
  # p < 0.05, log2fc > 0 -------------------------------------------------------
  vTopTable1 <-
    topTable( 
      veBayesFit,  
      coef = coef,  
      n = Inf, 
      p.value = 0.05,
      lfc = 0
    )
  path <- paste("../../results/", folder, model, "/",title,"_FDRq_0.05.tsv", sep = "") 
  write.table(
    vTopTable1,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )

  # p < 0.10, log2fc > 0 -------------------------------------------------------
  vTopTable2 <-
    topTable( 
      veBayesFit,  
      coef = coef,  
      n = Inf, 
      p.value = 0.1,
      lfc = 0
    )
  path <- paste("../../results/", folder, model, "/",title,"_FDRq_0.10.tsv", sep = "") 
  write.table(
    vTopTable2,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  
  # increment -----------------------------------------------------------------
  coef <- coef + 1
}
```

## Volcanos
```{r volcano_plot, message=FALSE, eval=FALSE}
# set variables
thresh <- 0.1

# get file list
files <- list.files(paste0("../../results/",folder,model,"/"))
keep <- grep("_1.00.tsv", files)
files <- files[keep]
files <- paste0("../../results/",folder,model,"/",files)

for (i in files) {
  
  # read DEG file
  data <- read.delim(i,sep = "\t")
  
  # assign colors
  color_values <- vector()
  max <- nrow(data)
  for(row in 1:max){
    if (data$adj.P.Val[row] < thresh){
      if (data$logFC [row] > 0){
        color_values <- c(color_values, 1) # 1 when logFC > 0 and FDRq < thresh
      }
      else if (data$logFC[row] < 0){
        color_values <- c(color_values, 2) # 2 when logFC < 0 and FDRq < thresh
      }
    }
    else{
      color_values <- c(color_values, 3) # 3 when FDRq >= thresh
    }
  }
  data$color_adjpval <- factor(color_values)
  
 # comparison name
 comparison <- gsub(paste0("../../results/",folder,model,"/"), "", i)
 comparison <- gsub("_FDRq_1.00.tsv","",comparison)
    
 # plot only if there are DEGs with adj.P.Val < thresh
 num <- subset(data, adj.P.Val < thresh)
 num <- nrow(num)
 if(num != 0) {
   
    # subset genes to label
    up <- data[data$color_adjpval == 1,]
    up15 <- up[1:15,]
    down <- data[data$color_adjpval == 2,]
    down15 <- down[1:15,]
      
    # set manual colors
    if (!1 %in% unique(data$color_adjpval)) {
      my_colors <- c("blue","gray")
    } else if (!2 %in% unique(data$color_adjpval)) {
      my_colors <- c("red","gray")
    } else if (!1 %in% unique(data$color_adjpval) && !2 %in% unique(data$color_adjpval)) {
      my_colors <- c("gray")
    } else {
      my_colors <- c("red","blue","gray")
    }
      
    # set significance threshold
    hadjpval <- (-log10(max(data$P.Value[data$adj.P.Val < thresh], na.rm=TRUE)))

    # plot
    p <-
      ggplot(data = data, 
             aes(x = logFC,  # x-axis is logFC
                 y = -log10(P.Value),  # y-axis will be -log10 of P.Value
                 color = color_adjpval)) +  # color is based on factored color column
      geom_point(alpha = 0.8, size = 2) +  # create scatterplot, alpha makes points transparent
      theme_bw() +  # set color theme
      theme(legend.position = "none") +  # no legend
      scale_color_manual(values = my_colors) +  # set factor colors
      labs(
        title = "", # no main title
        x = expression(log[2](FC)), # x-axis title
         y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)") # y-axis title
      ) +
      theme(axis.title.x = element_text(size = 15),
            axis.text.x = element_text(size = 15)) +
      theme(axis.title.y = element_text(size = 15),
            axis.text.y = element_text(size = 15)) +
      theme(plot.title = element_text(size = 15)) +
      geom_hline(yintercept = hadjpval,  #  horizontal line
                         colour = "#000000",
                         linetype = "dashed") +
      ggtitle(paste0(comparison, ", adj.P.Val < ", thresh)) +
      geom_text_repel(data = up15,
                      aes(x = logFC, y= -log10(P.Value), label = gene_name), 
                      size = 5,
                      color = "maroon", 
                      fontface="italic",
                      max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                      ) +
      geom_text_repel(data = down15,
                      aes(x = logFC, y= -log10(P.Value), label = gene_name), 
                      color = "navyblue", 
                      size = 5,
                      fontface="italic",
                      max.overlaps = getOption("ggrepel.max.overlaps", default = 30)
                      )
     p
      
    # save
    path <- paste0("../../results/",folder, model,"/volcano_FDRq_", format(thresh, nsmall = 2), ".pdf")
    pdf(path, height = 8, width = 8)
    print(p)
    dev.off()
  }
} # end loop through variables
```

```{r}
sessionInfo()
```
